anonymizer/OpenDiffix.Service/Led.fs

```diff
@@ -0,0 +1,225 @@
+  let groupingLabelsLength = aggregationContext.GroupingLabels.Length
+
+  let nonIsolatingColumns =
+    [| 0 .. groupingLabelsLength - 1 |]
+    |> Array.filter (fun column -> isolatingColumns |> Array.contains column |> not)
+
+  let anonAggregators = anonymizingAggregatorIndexes aggregationContext
+
+  let isolatorCache = Dictionary<Row, MutableList<Bucket * bool>>(Row.equalityComparer)
+
+  let victimBuckets =
+    buckets
+    |> Array.choose (fun bucket ->
+      let lowCount = Utils.isLowCount lowCountIndex bucket
+      let cacheKey = nonIsolatingColumns |> Array.map (fun i -> bucket.Group.[i])
+      let siblings = isolatorCache |> Dictionary.getOrInit cacheKey (fun _ -> MutableList())
+      siblings.Add((bucket, lowCount))
+      if lowCount then bucketsLowCount <- bucketsLowCount + 1
+      if lowCount && bucket.RowCount <= 3 then Some(bucket, cacheKey) else None
+    )
+
+  // Main victim bucket loop
+  for victimBucket, cacheKey in victimBuckets do
+    let victimGroup = victimBucket.Group
+    let siblingsPerColumn = Array.create groupingLabelsLength NoBuckets
+    let isolatorCandidates = isolatorCache.[cacheKey]
+
+    // Test for isolating columns
+    for otherBucket, otherLowCount in isolatorCandidates do
+      if not (obj.ReferenceEquals(victimBucket, otherBucket)) then
+        comparisons <- comparisons + 1UL
+
+        match findSingleNonMatchingColumn groupingLabelsLength victimGroup otherBucket.Group with
+        | Some nonMatchingColumn ->
+          siblingsPerColumn.[nonMatchingColumn] <-
+            match siblingsPerColumn.[nonMatchingColumn] with
+            | NoBuckets -> SingleBucket(otherBucket, otherLowCount)
+            | _ -> MultipleBuckets
+        | None -> ()
+
+    let hasIsolatingColumn =
+      siblingsPerColumn
+      |> Array.exists (
+        function
+        | SingleBucket (_, false) -> true
+        | _ -> false
+      )
+
+    if hasIsolatingColumn then
+      bucketsIsolatingColumn <- bucketsIsolatingColumn + 1
+
+      // todo, unknown cols
+
+      if hasUnknownColumn siblingsPerColumn then
+        bucketsMerged <- bucketsMerged + 1
+
+        siblingsPerColumn
+        |> Array.iter (
+          function
+          | SingleBucket (siblingBucket, false) ->
+            bumpDistribution victimBucket.RowCount
+            victimBucket |> mergeGivenAggregatorsInto siblingBucket anonAggregators
+
+            victimBucket
+            |> Bucket.putAttribute BucketAttributes.IS_LED_MERGED (Boolean true)
+          | _ -> ()
+        )
+
+  let isolatingColumnsStr = isolatingColumns |> Seq.map string |> String.join ", "
+
+  let mergeHistogramStr =
+    mergeHistogram
+    |> Seq.sortBy (fun pair -> pair.Key)
+    |> Seq.map (fun pair -> $"({pair.Key}, {pair.Value})")
+    |> Seq.append [ $"(*, {mergeHistogram |> Seq.sumBy (fun pair -> pair.Value)})" ]
+    |> String.join " "
+
+  Logger.debug $"[LED] Total buckets: {buckets.Length}"
```

Comment:
maybe it would be worth to extract everything diagnostic-related to a `Led.Diagnostics` module to make it stand out visually and make following the algo easier?

Reply:
Good idea! A class or something which holds this state could work well